### ISSUE TITLE
[QUERY] Service Bus client - implicit batching

### ISSUE THREAD

**Issue by starkpl** [Azure/azure-sdk-for-net#27411]
### Library name and version

Azure.Messaging.ServiceBus 7.5.0

### Query/Question

Is there an implicit batching mechanism built into Service Bus SDK? What I mean by implicit batching is that I'm calling `ServiceBusSender.SendMessageAsync` method multiple times in a short interval and they are being somehow batched together before being sent to the Service Bus.

[The documentation](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-performance-improvements?tabs=net-standard-sdk-2#batching-store-access) states that there is a batching mechanism on the server side of the Service Bus. However, I run a set of tests that shows there probably is some kind of batching even when the server-side "batching store access" is disabled on the topic. What's more, I observed that when sending each message from a separate SDK sender instance the batching stopped occurring, which may suggest that the batching is occurring inside the SDK sender (client-side). 

The assumption that the batching is happening is based on the observed Incoming Requests metric on the Service Bus side and its comparison to the Incoming Messages metric. When reusing a single SDK sender, the value of the Incoming Requests metric is much lower than the Incoming Messages, but when using different SDK senders for each message, the metric values are equal.

For testing I used a performance benchmark tool created by Microsoft. The exact configuration I used and test results are available in my forked version of this tool (link below), so it can be easily reproduced. To limit the machine's resource utilization, instead of creating an SDK sender for each message I actually used a pool of 1.000 senders created before the test.
https://github.com/starkpl/service-bus-dotnet-messaging-performance/tree/master/ThroughputTest#sdk-senders-pooling-experiment

I would be very grateful for a further explanation of how exactly does implicit batching work in the Service Bus, because it would be very useful when deciding on the topology of topics and for proper usage of the SDK.

**Comment by jsquire** [comment]
Hi @starkpl.   Thank you for reaching out; if I understand correctly, what you're asking about is the implicit batching behavior that the Service Bus service applies when persisting messages to its durable storage.   I've routed this over to the Service Bus team, who is best positioned to assist with those answers.

**Comment by starkpl** [comment]
Thanks, @jsquire. Maybe I will try to summarize what exactly is my point (and question):
- I've observed that there is some kind of implicit batching when sending messages to a topic because I can see that the Incoming Requests value is much lower than Incoming Messages.
- This is still happening even if I disable the server-side "batching store access" on the topic ([documented here](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-performance-improvements?tabs=net-standard-sdk-2#batching-store-access)). So this mechanism is not responsible for the batching I've observed.
- This stops when I use a large number of SDK senders (so I no longer reuse a single SDK sender instance). In this case, the Incoming Requests equals the Incoming Messages. This may suggest that the implicit batching is done client-side, inside SDK sender, but it's just my assumption.
- It can be easily reproduced using [this tool](https://github.com/starkpl/service-bus-dotnet-messaging-performance/tree/master/ThroughputTest#sdk-senders-pooling-experiment) (fork of MS's benchmark tool).

To summarize my question: is this implicit batching that I've observed happening on the client-side or the server-side? And is there any documentation about it?

**Comment by jsquire** [comment]
> happening on the client-side

I can confirm that there is no implicit batching happening on the client-side.   Batches are explicitly managed by callers via the `ServiceBusMessageBatch` abstraction.   Any calls to `SendMessageAsync` are treated as individual operations.

**Comment by starkpl** [comment]
Thank you, so I will wait for the input from the Service Bus team - maybe they can explain this behaviour or point me to the documentation.

**Comment by danielmarbach** [comment]
@jsquire didn't the Amqp lib use a disposition interval at some point that batched things on the same session together?

https://github.com/Azure/azure-amqp/blob/59894e8d992c63dfcaad6d1d3f454d9d88855d91/src/AmqpSessionSettings.cs#L40

Ah it looks like the previous value of 20 ms got removed at some point on master. 

https://github.com/Azure/azure-amqp/commit/82fdcf7c00f7c9671d3a613064f8f082b001df4e

The question is though if it is in the released version...

**Comment by jsquire** [comment]
Apologies; I missed the callout from @danielmarbach.   I wasn't aware of that setting ever being populated.   My understanding from double-checking with Xin lines up with his feedback in the issue that @starkpl linked; there was no implicit batching behavior even when it was set.

**Comment by EldertGrootenboer** [comment]
We do batch messages when [writing to the internal store ](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-performance-improvements?tabs=net-standard-sdk-2#batching-store-access) as you found, but other then that there is no batching on the service side.

**Comment by starkpl** [comment]
Thank you @EldertGrootenboer. Just to make sure, is this batching mechanism supposed to work only "per client"? Meaning that if I have different clients (let's say different processes on different machines) sending messages on exactly the same topic, they will not be batched together by design (only messages from the same client would be batched)? I am asking about this because in distributed scenarios (multiple senders distributed on multiple nodes etc) it would be very beneficial if messages from different clients could be batched together - otherwise impact of this feature may be limited.

Also in my experimental repo, I showed how disabling batched store access on the topic didn't change the batching I have observed using metrics. Is it possible that the `EnableBatchOperations` flag on the topic is no longer working? It can be reproduced as described here:
https://github.com/starkpl/service-bus-dotnet-messaging-performance/tree/master/ThroughputTest#no-pooling-scenario-with-disabled-batched-store-access

**Comment by EldertGrootenboer** [comment]
We have added an investigation task for this, and will update this issue when we have more information.

**Comment by EldertGrootenboer** [comment]
We have reassigned an engineer to look further into this issue. We will update this issue when we have more information.

**Comment by benfoster** [comment]
Any update on this?

**Comment by yvgopal** [comment]
Yes. There is implicit batching of incoming messages on each AMQP send link. It doesn't batch messages across links. If there are multiple messages arriving on a link, we internally do batch them for efficiency. We also do further batching when writing to store, which depends on the store again. I think you shouldn't make assumptions about our internal batching logic when designing your application. Our batching logic may change anytime in the future, if there is a positive trade off.

EnableBatchedOperations flag doesn't influence this behavior. In fact, EnableBatchedOperations is largely irrelevant now, except for one corner case. We will update the documentation about that property.
